src/main/java/org/truffleruby/language/methods/AddMethodNode.java

```diff
@@ -44,26 +43,17 @@ protected void addMethod(DynamicObject module, InternalMethod method, Visibility
         if (!ignoreNameVisibility && ModuleOperations.isMethodPrivateFromName(method.getName())) {
             visibility = Visibility.PRIVATE;
         }
-
         if (Layouts.MODULE.getFields(module).isRefinement()) {
             final DynamicObject refinedModule = Layouts.MODULE.getFields(module).getRefinedModule();
-            addRefinedMethodEntry(refinedModule, method, visibility);
+            handleRefinedMethodEntry(refinedModule, method);
```

Comment:
Should we call `handleRefinedMethod()` also when include/prepend is used on a refinement module (R)?

Reply:
The good point:

```ruby
module A
  def /(other) quo(other) end
end

module IntegerSlashExt
  refine Integer do
    include A
  end
end

using IntegerSlashExt

puts 1 / 2

# (MRI) Rational (1 / 2)
# (TR) 0 
```

yeah, this is because we don't invalidate `assumptions` when included\prepended are executed for R. 

I found the method is called for `prepend` and `include` and does the same sort of logic: 
https://github.com/oracle/truffleruby/blob/443e77af313d6b550e17398c0f31b6277ee69672/src/main/java/org/truffleruby/core/module/ModuleFields.java#L627-L630

I'm thinking about changing it to:

```java
public void newHierarchyVersion() {
    newConstantsVersion();
    newMethodsVersion();
    newBultinsVersion();
}

private void newBultinsVersion() {
    if (isRefinement()) {
        Layouts.MODULE.getFields(getRefinedModule()).invalidateBuiltinsAssumptions();
    }
}

- # private void prependInvalidation() {
+ public void invalidateBuiltinsAssumptions() {
    if (!inlinedBuiltinsAssumptions.isEmpty()) {
        for (Assumption assumption : inlinedBuiltinsAssumptions.values()) {
            assumption.invalidate();
        }
    }
}
```

I've checked and it works. And for sure we need to add a test for this case to the `ruby/spec` suite.

Reply:
That sounds like a nice way to deal with that :smiley: 
Yes, please add a spec, it's fine in `ruby/spec`, such invalidation must also exist in some form in other Ruby implementations (and anyway I'd say almost anything that runs on MRI and on some other Ruby implementation should go in ruby/spec, in order to avoid duplicate efforts for testing).